File: src/main/com/stockcomp/controller/common/CustomExceptionHandler.java
package com.stockcomp.controller.common;

import com.stockcomp.exception.InsufficientFundsException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.NoSuchElementException;

@ControllerAdvice
public class CustomExceptionHandler extends ResponseEntityExceptionHandler {

    @ExceptionHandler(NoSuchElementException.class)
    public ResponseEntity<Object> handleNoSuchElementException(NoSuchElementException exception) {
        return new ResponseEntity<>(exception.getMessage(), HttpStatus.NOT_FOUND);
    }

    @ExceptionHandler(InsufficientFundsException.class)
    public ResponseEntity<Object> handleInsufficientFundsException(InsufficientFundsException exception){
        return new ResponseEntity<>(exception.getMessage(), HttpStatus.CONFLICT);
    }
}

File: src/main/com/stockcomp/controller/common/CustomExceptionHandler.kt
package com.stockcomp.controller.common

import com.stockcomp.exception.InsufficientFundsException
import io.jsonwebtoken.ExpiredJwtException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

@RestControllerAdvice
class CustomExceptionHandler : ResponseEntityExceptionHandler() {

    @ExceptionHandler(NoSuchElementException::class)
    fun handleNoSuchElementException(exception: NoSuchElementException): ResponseEntity<Any> =
        ResponseEntity(exception.message, HttpStatus.NOT_FOUND)

    @ExceptionHandler(InsufficientFundsException::class)
    fun handleInsufficientFundsException(exception: InsufficientFundsException): ResponseEntity<Any> =
        ResponseEntity(exception.message, HttpStatus.CONFLICT)

    @ExceptionHandler(ExpiredJwtException::class)
    fun handleJwtExpiredException(exception: ExpiredJwtException): ResponseEntity<Any> =
        ResponseEntity("JWT Token expired", HttpStatus.UNAUTHORIZED)
}